README.md
# CF Post Formats

This is my fork of [CF Post Formats](https://github.com/crowdfavorite/wp-post-formats), a WordPress plugin that improves on the default post format UI.

* [Alex King's post about this project](http://alexking.org/blog/2011/10/25/wordpress-post-formats-admin-ui).
* [Original ticket on the WordPress trac](http://core.trac.wordpress.org/ticket/19570).
* [Post formats UI exist WordPress 3.6 core](http://make.wordpress.org/core/2013/05/29/post-formats-ui-is-exiting-core-will-live-as-a-plugin/).
* Merged with version 1.3.1 of the original plugin.
* Compatible with WordPress 3.8.1.

Modifications to the original plugin in this fork:

* Title and date fields added to the quotation post format UI to [allow compliance with the HTML5 specification](http://synapticism.com/working-with-wordpress-post-formats-and-quotations/).
* Autotitle feature now based on 10 words rather than 50 characters.
* Autotitle feature currently requires saving a draft before the slug is created.


CHANGELOG.txt
# CF Post Formats

## Version 1.4

- (fix) Gallery preview images are served over SSL if the admin page is loaded over SSL
- (fix) Suppress PHP warning that could display in some situations when loading gallery preview images
- (fix) jQuery live() is deprecated

## Version 1.3.1

- (fix) WordPress 3.8 admin UI compatibility

## Version 1.3

- (fix) WordPress 3.8 admin UI compatibility

## Version 1.2

- (fix) Don't auto-set the post title for status posts if a title already exists

## Version 1.1

- WordPress 3.5 compatibility
- Requires WordPress 3.5

## Version 1.0.2

- Gallery is displayed as expected when images have already been uploaded before gallery tab is selected
- Fix audio fields showing on Video tab (props DrewAPicture)
- Fix quote fields showing on Link tab (props dariodev)
- Compatibility with Social broadcasting of Status posts

## Version 1.0.1

- added CHANGELOG
- add compatibility with `pre_ping` changes in WordPress 3.4


## Version 1.0

- first public release


js/admin.js
jQuery(function($) {
	var CF = CF || {};
	
	CF.postFormats = function($) {
		return {
			switchTab: function(clicked) {
				var $this = $(clicked),
					$tab = $this.closest('li');

				if (!$this.hasClass('current')) {
					$this.addClass('current');
					$tab.siblings().find('a').removeClass('current');
					this.switchWPFormat($this.attr('href'));
				}
			},
			
			switchWPFormat: function(formatHash) {
				$(formatHash).trigger('click');
				switch (formatHash) {
					case '#post-format-0':
					case '#post-format-aside':
					case '#post-format-chat':
						CF.postFormats.standard();
						break;
					case '#post-format-status':
					case '#post-format-link':
					case '#post-format-image':
					case '#post-format-gallery':
					case '#post-format-video':
					case '#post-format-quote':
					case '#post-format-audio':
						CF.postFormats[formatHash.replace('#post-format-', '')]();
				}
				$(document).trigger('cf-post-formats-switch', formatHash);
			},

			standard: function() {
				$('#cfpf-format-link-url, #cfpf-format-quote-fields, #cfpf-format-video-fields, #cfpf-format-audio-fields, #cfpf-format-gallery-preview').hide();
				$('#titlewrap').show();
				$('#postimagediv-placeholder').replaceWith($('#postimagediv'));
			},
			
			status: function() {
				$('#titlewrap, #cfpf-format-link-url, #cfpf-format-quote-fields, #cfpf-format-video-fields, #cfpf-format-audio-fields, #cfpf-format-gallery-preview').hide();
				$('#postimagediv-placeholder').replaceWith($('#postimagediv'));
				$('#content:visible').focus();
			},

			link: function() {
				$('#cfpf-format-quote-fields, #cfpf-format-video-fields, #cfpf-format-audio-fields, #cfpf-format-gallery-preview').hide();
				$('#titlewrap, #cfpf-format-link-url').show();
				$('#postimagediv-placeholder').replaceWith($('#postimagediv'));
			},
			
			image: function() {
				$('#cfpf-format-link-url, #cfpf-format-quote-fields, #cfpf-format-video-fields, #cfpf-format-audio-fields, #cfpf-format-gallery-preview').hide();
				$('#titlewrap').show();
				$('#postimagediv').after('<div id="postimagediv-placeholder"></div>').insertAfter('#titlediv');
			},

			gallery: function() {
				$('#cfpf-format-link-url, #cfpf-format-quote-fields, #cfpf-format-video-fields, #cfpf-format-audio-fields').hide();
				$('#titlewrap, #cfpf-format-gallery-preview').show();
				$('#postimagediv-placeholder').replaceWith($('#postimagediv'));
			},

			video: function() {
				$('#cfpf-format-link-url, #cfpf-format-quote-fields, #cfpf-format-gallery-preview, #cfpf-format-audio-fields').hide();
				$('#titlewrap, #cfpf-format-video-fields').show();
				$('#postimagediv-placeholder').replaceWith($('#postimagediv'));
			},

			quote: function() {
				$('#titlewrap, #cfpf-format-link-url, #cfpf-format-video-fields, #cfpf-format-audio-fields, #cfpf-format-gallery-preview').hide();
				$('#cfpf-format-quote-fields').show().find(':input:first').focus();
				$('#postimagediv-placeholder').replaceWith($('#postimagediv'));
			},

			audio: function() {
				$('#cfpf-format-link-url, #cfpf-format-quote-fields, #cfpf-format-video-fields, #cfpf-format-gallery-preview').hide();
				$('#titlewrap, #cfpf-format-audio-fields').show();
				$('#postimagediv-placeholder').replaceWith($('#postimagediv'));
			},

			gallerySortable: function() {
				$galleryPreview = $('#cfpf-format-gallery-preview .gallery');
				$galleryPreview.sortable({
					containment: 'parent',
					cursor: 'move',
					forceHelperSize: true,
					forcePlaceholderSize: true,
					update: function() {
						var ids = [];
						$galleryPreview.find('img.attachment-thumbnail').each(function() {
							ids.push($(this).data('id'));
						});

						$.post(
							ajaxurl,
							{
								'action': 'cfpf_gallery_menu_order',
								'order': ids
							}
						);
					}
				});
			}
		};
	}(jQuery);
	
	// move tabs in to place
	$('#cf-post-format-tabs').insertBefore($('form#post')).show();
	$('#cfpf-format-link-url, #cfpf-format-video-fields, #cfpf-format-audio-fields').insertAfter($('#titlediv'));
	$('#cfpf-format-gallery-preview').find('dt a').each(function() {
		$(this).replaceWith($(this.childNodes)); // remove links
	}).end().insertAfter($('#titlediv'));
	$('#cfpf-format-quote-fields').insertAfter($('#titlediv'));
	
	$(document).trigger('cf-post-formats-init');
	
	// tab switch
	$(document).on('click', '#cf-post-format-tabs a', function(e) {
		CF.postFormats.switchTab(this);
		e.stopPropagation();
		e.preventDefault();
	}).filter('.current').each(function() {
		CF.postFormats.switchWPFormat($(this).attr('href'));
	});
	
	// set to shortcode when clicking into shortcode field
	$(document).on('click focus', '#cfpf-format-gallery-shortcode', function() {
		$('#cfpf-format-gallery-type-shortcode').prop('checked', true);
	});

	// WordPress 3.5 compatibility
	// props: https://gist.github.com/4192094
	
	var postId = $('#post_ID').val();
	var gallery = wp.media.query({ uploadedTo: postId });

	// Run the query.
	// This returns a promise (like $.ajax) so you can do things when it completes.
	gallery.more();

	// Bind your events for when the contents of the gallery changes.
	gallery.on( 'add remove reset', function() {
		// Something changed, update your stuff.

		var $preview = $('#cfpf-format-gallery-preview');
// spinner
		$preview.find('.cf-elm-container').html('<p><img src="' + cfpf_post_format.wpspin_light + '" alt="' + cfpf_post_format.loading + '" /></p>');
// AJAX call for gallery snippet
		$.post(
			ajaxurl,
			{
				'action': 'cfpf_gallery_preview',
				'id': $('#post_ID').val()
			},
			function(response) {
// replace
				$preview.replaceWith(response.html);
// find it again
				$preview = $('#cfpf-format-gallery-preview')
				$preview.find('dt a').each(function() {
					$(this).replaceWith($(this.childNodes)); // remove links
				}).end();
// only show if tab is selected
				if ($('#cf-post-format-tabs a.current').attr('href').indexOf('#post-format-gallery') != -1) {
					$preview.show();
				}

				CF.postFormats.gallerySortable();
			},
			'json'
		);

	}, gallery );

	$(document).on('click', '#cfpf-format-gallery-preview .none a', function(e) {
		$('#wp-content-media-buttons .insert-media').mousedown().mouseup().click();
		e.preventDefault();
	});

	CF.postFormats.gallerySortable();
});


views/format-quote.php
<div id="cfpf-format-quote-fields" style="display: none;">
	<div class="cf-elm-block">
		<label for="cfpf-format-quote-source-name"><?php _e('Source Name', 'cf-post-format'); ?></label>
		<input type="text" name="_format_quote_source_name" value="<?php echo esc_attr(get_post_meta($post->ID, '_format_quote_source_name', true)); ?>" id="cfpf-format-quote-source-name" tabindex="1" />
	</div>
	<div class="cf-elm-block">
		<label for="cfpf-format-quote-source-name"><?php _e('Source Title', 'cf-post-format'); ?></label>
		<input type="text" name="_format_quote_source_title" value="<?php echo esc_attr(get_post_meta($post->ID, '_format_quote_source_title', true)); ?>" id="cfpf-format-quote-source-name" tabindex="1" />
	</div>
	<div class="cf-elm-block">
		<label for="cfpf-format-quote-source-name"><?php _e('Source Date', 'cf-post-format'); ?></label>
		<input type="text" name="_format_quote_source_date" value="<?php echo esc_attr(get_post_meta($post->ID, '_format_quote_source_date', true)); ?>" id="cfpf-format-quote-source-name" tabindex="1" />
	</div>
	<div class="cf-elm-block">
		<label for="cfpf-format-quote-source-url"><?php _e('Source URL', 'cf-post-format'); ?></label>
		<input type="text" name="_format_quote_source_url" value="<?php echo esc_attr(get_post_meta($post->ID, '_format_quote_source_url', true)); ?>" id="cfpf-format-quote-source-url" tabindex="1" />
	</div>
</div>

views/format-gallery.php
<?php

$gallery_type = cfpf_post_gallery_type();

?>
<div id="cfpf-format-gallery-preview" class="cf-elm-block cf-elm-block-image" style="display: none;">

	<label><span><?php _e('Gallery Images', 'cf-post-format'); ?></span></label>

	<div class="cf-elm-container cfpf-gallery-options">
		<p>
			<input type="radio" name="_format_gallery_type" value="shortcode" <?php checked($gallery_type, 'shortcode' ); ?> id="cfpf-format-gallery-type-shortcode"  />
			<label for="cfpf-format-gallery-type-shortcode"><?php _e('Shortcode', 'cf-post-format'); ?></label>
			<input type="text" name="_format_gallery_shortcode" value="<?php echo esc_attr(get_post_meta($post->ID, '_format_gallery_shortcode', true)); ?>" id="cfpf-format-gallery-shortcode" />
		</p>
		<p>
			<input type="radio" name="_format_gallery_type" value="attached-images" <?php checked($gallery_type, 'attached-images' ); ?> id="cfpf-format-gallery-type-attached" />
			<label for="cfpf-format-gallery-type-attached"><?php _e('Images uploaded to this post', 'cf-post-format'); ?></label>
		</p>


<?php

// running this in the view so it can be used by multiple functions

$attachments = get_posts(array(
	'post_type' => 'attachment',
	'numberposts' => -1,
	'post_status' => null,
	'post_parent' => $post->ID,
	'order' => 'ASC',
	'orderby' => 'menu_order ID',
));

if ($attachments) {

	if (is_ssl()) {
		add_filter('wp_get_attachment_image_attributes', 'cfpf_ssl_gallery_preview', 10, 2);
	}

	echo '<ul class="gallery clearfix">';
	foreach ($attachments as $attachment) {
		echo '<li>'.wp_get_attachment_image($attachment->ID, 'thumbnail').'</li>';
	}
	echo '</ul>';
}

?>
		<p class="none" style="float: none; clear: both;">
			<a href="#" class="button"><?php _e('Upload Images', 'cf-post-format'); ?></a>
		</p>
	</div>
</div>


cf-post-formats.php
<?php
/*
Plugin Name: CF Post Formats
Plugin URI: http://crowdfavorite.com
Description: Custom post format admin UI
Text Domain: cf-post-format
Domain Path: /languages/
Version: 1.3.1
Author: crowdfavorite
Author URI: http://crowdfavorite.com
*/

/**
 * Copyright (c) 2011-2013 Crowd Favorite, Ltd. All rights reserved.
 *
 * Released under the GPL license
 * http://www.opensource.org/licenses/gpl-license.php
 *
 * This is an add-on for WordPress
 * http://wordpress.org/
 *
 * **********************************************************************
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * **********************************************************************
 */

if (!defined('CFPF_VERSION')) {

define('CFPF_VERSION', '1.3');

function cfpf_base_url() {
	return trailingslashit(apply_filters('cfpf_base_url', plugins_url('', __FILE__)));
}

/**
 * Load translation files
 */
function cf_post_format_translation_files() {
  $theload = load_plugin_textdomain( 'cf-post-format', false, dirname( plugin_basename( __FILE__ ) ) . '/languages/' );
 }
 
add_action( 'init', 'cf_post_format_translation_files' );

function cfpf_admin_init() {
	$post_formats = get_theme_support('post-formats');
	if (!empty($post_formats[0]) && is_array($post_formats[0])) {
		if (in_array('link', $post_formats[0])) {
			add_action('save_post', 'cfpf_format_link_save_post');
		}
		if (in_array('status', $post_formats[0])) {
			add_action('save_post', 'cfpf_format_status_save_post', 10, 2);
		}
		if (in_array('quote', $post_formats[0])) {
			add_action('save_post', 'cfpf_format_quote_save_post', 10, 2);
		}
		if (in_array('video', $post_formats[0])) {
			add_action('save_post', 'cfpf_format_video_save_post');
		}
		if (in_array('audio', $post_formats[0])) {
			add_action('save_post', 'cfpf_format_audio_save_post');
		}
		if (in_array('gallery', $post_formats[0])) {
			add_action('save_post', 'cfpf_format_gallery_save_post');
		}
	}
}
add_action('admin_init', 'cfpf_admin_init');

// we aren't really adding meta boxes here,
// but this gives us the info we need to get our stuff in.
function cfpf_add_meta_boxes($post_type) {
	if (post_type_supports($post_type, 'post-formats') && current_theme_supports('post-formats')) {
		// assets
		wp_enqueue_script('cf-post-formats', cfpf_base_url().'js/admin.js', array('jquery'), CFPF_VERSION);
		wp_enqueue_style('cf-post-formats', cfpf_base_url().'css/admin.css', array(), CFPF_VERSION, 'screen');

		wp_localize_script(
			'cf-post-formats',
			'cfpf_post_format',
			array(
				'loading' => __('Loading...', 'cf-post-formats'),
				'wpspin_light' => admin_url('images/wpspin_light.gif')
			)
		);

		add_action('edit_form_after_title', 'cfpf_post_admin_setup');
	}
}
add_action('add_meta_boxes', 'cfpf_add_meta_boxes');

/**
 * Show the post format navigation tabs
 * A lot of cues are taken from the `post_format_meta_box`
 *
 * @return void
 */
function cfpf_post_admin_setup() {
	$post_formats = get_theme_support('post-formats');
	if (!empty($post_formats[0]) && is_array($post_formats[0])) {
		global $post;
		$current_post_format = get_post_format($post->ID);

		// support the possibility of people having hacked in custom
		// post-formats or that this theme doesn't natively support
		// the post-format in the current post - a tab will be added
		// for this format but the default WP post UI will be shown ~sp
		if (!empty($current_post_format) && !in_array($current_post_format, $post_formats[0])) {
			array_push($post_formats[0], get_post_format_string($current_post_format));
		}
		array_unshift($post_formats[0], 'standard');
		$post_formats = $post_formats[0];

		include('views/tabs.php');

		$format_views = array(
			'link',
			'quote',
			'video',
			'gallery',
			'audio',
		);
		foreach ($format_views as $format) {
			if (in_array($format, $post_formats)) {
				include('views/format-'.$format.'.php');
			}
		}
	}
}

function cfpf_format_link_save_post($post_id) {
	if (!defined('XMLRPC_REQUEST') && isset($_POST['_format_link_url'])) {
		update_post_meta($post_id, '_format_link_url', $_POST['_format_link_url']);
	}
}
// action added in cfpf_admin_init()

function cfpf_format_auto_title_post($post_id, $post) {
	// get out early if a title is already set
	if (!empty($post->post_title)) {
		return;
	}

	remove_action('save_post', 'cfpf_format_status_save_post', 10, 2);
	remove_action('save_post', 'cfpf_format_quote_save_post', 10, 2);

	$content = trim(strip_tags($post->post_content));
	$title = substr($content, 0, 50);
	if (strlen($content) > 50) {
		$title .= '...';
	}
	$title = apply_filters('cfpf_format_auto_title', $title, $post);
	wp_update_post(array(
		'ID' => $post_id,
		'post_title' => $title
	));

	add_action('save_post', 'cfpf_format_status_save_post', 10, 2);
	add_action('save_post', 'cfpf_format_quote_save_post', 10, 2);
}

function cfpf_format_status_save_post($post_id, $post) {
	if (has_post_format('status', $post)) {
		cfpf_format_auto_title_post($post_id, $post);
	}
}
// action added in cfpf_admin_init()

function cfpf_format_quote_save_post($post_id, $post) {
	if (!defined('XMLRPC_REQUEST')) {
		$keys = array(
			'_format_quote_source_name',
			'_format_quote_source_url',
		);
		foreach ($keys as $key) {
			if (isset($_POST[$key])) {
				update_post_meta($post_id, $key, $_POST[$key]);
			}
		}
	}
	if (has_post_format('quote', $post)) {
		cfpf_format_auto_title_post($post_id, $post);
	}
}
// action added in cfpf_admin_init()

function cfpf_format_video_save_post($post_id) {
	if (!defined('XMLRPC_REQUEST') && isset($_POST['_format_video_embed'])) {
		update_post_meta($post_id, '_format_video_embed', $_POST['_format_video_embed']);
	}
}
// action added in cfpf_admin_init()

function cfpf_format_audio_save_post($post_id) {
	if (!defined('XMLRPC_REQUEST') && isset($_POST['_format_audio_embed'])) {
		update_post_meta($post_id, '_format_audio_embed', $_POST['_format_audio_embed']);
	}
}
// action added in cfpf_admin_init()

/**
 * Updates the _format_gallery values in the DB for
 * the radio buttons and text field in the gallery format tab.
 *
 *
 * @param int $post_id The id of the post.
 * @return void
 */
function cfpf_format_gallery_save_post( $post_id ) {
	if (!defined('XMLRPC_REQUEST')) {
		$keys = array(
			'_format_gallery_shortcode',
			'_format_gallery_type'
		);
		foreach ($keys as $key) {
			if (isset($_POST[$key])) {
				update_post_meta($post_id, $key, $_POST[$key]);
			}
		}
	}
}
// action added in cfpf_admin_init()

function cfpf_gallery_preview() {
	if (empty($_POST['id']) || !($post_id = intval($_POST['id']))) {
		exit;
	}

	global $post;
	$post = get_post($post_id);
	ob_start();
	include('views/format-gallery.php');
	$html = ob_get_clean();
	header('Content-type: text/javascript');
	echo json_encode(compact('html'));
	exit;
}
add_action('wp_ajax_cfpf_gallery_preview', 'cfpf_gallery_preview');

// filter added conditionally in views/format-gallery.php
function cfpf_ssl_gallery_preview($attr, $attachment) {
	$attr['src'] = str_replace('http://', 'https://', $attr['src']);
	return $attr;
}

function cfpf_post_has_gallery($post_id = null) {
	if (empty($post_id)) {
		$post_id = get_the_ID();
	}
	$images = new WP_Query(array(
		'post_parent' => $post_id,
		'post_type' => 'attachment',
		'post_status' => 'inherit',
		'posts_per_page' => 1, // -1 to show all
		'post_mime_type' => 'image%',
		'orderby' => 'menu_order',
		'order' => 'ASC'
	));
	return (bool) $images->post_count;
}

// ensure that we have expected data and set a default
// for backward compatibility.
function cfpf_post_gallery_type() {
	$post = get_post();
	$value = get_post_meta($post->ID, '_format_gallery_type', true);
	switch ($value) {
		case 'shortcode':
		case 'attached-images':
			$value = $value;
		break;
		default:
			$value = 'attached-images';
	}
	return $value;
}

// accepts an associative array of args to added to the shortcode before output
function cfpf_gallery_output($args = array()) {
	// setup args if we have any
	$args_string = '';
	if (!empty($args)) {
		foreach ($args as $k => $v) {
			$args_string .= $k.'="'.esc_attr($v).'" ';
		}
	}
	$type = cfpf_post_gallery_type();
	// if type is set to shortcode, make sure we have something as a shortcode
	if ($type == 'shortcode') {
		$post = get_post();
		$shortcode = trim(get_post_meta($post->ID, '_format_gallery_shortcode', true));
		if (!empty($shortcode) && substr($shortcode, -1) == ']') {
			// add args
			$shortcode = substr($shortcode, 0, -1).' '.$args_string.']';
			echo do_shortcode($shortcode);
			return;
		}
	}
	// or fall back to attached images
	echo do_shortcode('[gallery '.$args_string.']');
}

function cfpf_pre_ping_post_links($post_links, $pung, $post_id = null) {
	// return if we don't get a post ID (pre WP 3.4)
	if (empty($post_id)) {
		return;
	}
	$url = get_post_meta($post_id, '_format_link_url', true);
	if (!empty($url) && !in_array($url, $pung) && !in_array($url, $post_links)) {
		$post_links[] = $url;
	}
}
add_filter('pre_ping', 'cfpf_pre_ping_post_links', 10, 3);

// For integration with Social plugin (strips {title} from broadcast format on status posts)
function cfpf_social_broadcast_format($format, $post) {
	if (get_post_format($post) == 'status') {
		$format = trim(str_replace(
			array(
				'{title}:',
				'{title} -',
				'{title}',
			),
			'',
			$format
		));
	}
	return $format;
}
add_filter('social_broadcast_format', 'cfpf_social_broadcast_format', 10, 2);


} // end defined check
